web/src/pages/billing/components/HailBatchGrid.tsx

```diff
@@ -0,0 +1,498 @@
+                )
+                aggArGUIDData[idx].end_time = new Date(
+                    Math.max(usageEndDate.getTime(), aggArGUIDData[idx].end_time.getTime())
+                )
+            }
+        }
+    })
+
+    const aggArGUIDResource: any[] = []
+    data.forEach((curr) => {
+        const { cost, batch_resource } = curr
+        const ar_guid = curr['ar-guid']
+        const idx = aggArGUIDResource.findIndex(
+            (d) => d.ar_guid === ar_guid && d.batch_resource === batch_resource
+        )
+        if (cost >= 0) {
+            // do not include credits, should be filter out at API?
+            if (idx === -1) {
+                aggArGUIDResource.push({
+                    type: 'ar_guid',
+                    key: ar_guid,
+                    ar_guid,
+                    batch_resource,
+                    cost,
+                })
+            } else {
+                aggArGUIDResource[idx].cost += cost
+            }
+        }
+    })
+    const aggBatchData: any[] = []
+    data.forEach((curr) => {
+        const {
+            batch_id,
+            url,
+            topic,
+            namespace,
+            batch_name,
+            cost,
+            usage_start_time,
+            usage_end_time,
+        } = curr
+        const ar_guid = curr['ar-guid']
+        const usageStartDate = new Date(usage_start_time)
+        const usageEndDate = new Date(usage_end_time)
+        const idx = aggBatchData.findIndex(
+            (d) =>
+                d.batch_id === batch_id &&
+                d.batch_name === batch_name &&
+                d.topic === topic &&
+                d.namespace === namespace
+        )
+        if (cost >= 0) {
+            // do not include credits, should be filter out at API?
+            if (idx === -1) {
+                aggBatchData.push({
+                    type: 'batch_id',
+                    key: batch_id,
+                    ar_guid,
+                    batch_id,
+                    url,
+                    topic,
+                    namespace,
+                    batch_name,
+                    job_id: undefined,
+                    cost,
+                    start_time: usageStartDate,
+                    end_time: usageEndDate,
+                })
+            } else {
+                aggBatchData[idx].cost += cost
+                aggBatchData[idx].start_time = new Date(
+                    Math.min(usageStartDate.getTime(), aggBatchData[idx].start_time.getTime())
+                )
+                aggBatchData[idx].end_time = new Date(
+                    Math.max(usageEndDate.getTime(), aggBatchData[idx].end_time.getTime())
+                )
+            }
+        }
+    })
+
+    const aggBatchResource: any[] = []
+    data.forEach((curr) => {
+        const { batch_id, batch_resource, topic, namespace, batch_name, cost } = curr
+        const ar_guid = curr['ar-guid']
+        const idx = aggBatchResource.findIndex(
```

Comment:
How large is aggBatchResource? Can we key by different things to turn this sort O(n) (in an O(m) loop) to O(1)?

Reply:
Data are only small set, specific per one AR_GUID, so I do not think it is going to be a problem.
Happy to revisit if we encounter this as an issue.